src/block_gmres.jl

```diff
@@ -0,0 +1,349 @@
+# An implementation of block-GMRES for the solution of the square linear system AX = B.
+#
+# Alexis Montoison, <alexis.montoison@polymtl.ca>
+# Chicago, October 2023.
+
+include("block_utils.jl")
+
+export BlockGmresSolver, block_gmres, block_gmres!
+
+"""
+Type for storing the vectors required by the in-place version of BLOCK-GMRES.
+
+The outer constructors
+
+    solver = BlockGmresSolver(m, n, p, memory, S)
+    solver = BlockGmresSolver(A, B, memory=5)
+
+may be used in order to create these vectors.
+`memory` is set to `div(n,p)` if the value given is larger than `div(n,p)`.
+"""
+mutable struct BlockGmresSolver{T,FC,S} <: KrylovSolver{T,FC,S}
+  m          :: Int
+  n          :: Int
+  p          :: Int
+  ΔX         :: S
+  X          :: S
+  W          :: S
+  P          :: S
+  Q          :: S
+  V          :: Vector{S}
+  Z          :: Vector{S}
+  R          :: Vector{S}
+  warm_start :: Bool
+  stats      :: BlockGmresStats{T}
+end
+
+function BlockGmresSolver(m, n, p, memory, S)
+  memory = min(div(n,p), memory)
+  FC = eltype(S)
+  T  = real(FC)
+  ΔX = S(undef, 0, 0)
+  X  = S(undef, n, p)
+  W  = S(undef, n, p)
+  P  = S(undef, 0)
+  Q  = S(undef, 0)
+  V  = S[S(undef, n, p) for i = 1 : memory]
+  Z  = S[S(undef, p, p) for i = 1 : memory]
+  R  = S[S(undef, p, p) for i = 1 : div(memory * (memory+1), 2)]
+  stats = BlockGmresStats(0, false, T[], 0.0, "unknown")
+  solver = BlockGmresSolver{T,FC,S}(m, n, p, ΔX, X, W, P, Q, V, Z, R, false, stats)
+  return solver
+end
+
+function GmresSolver(A, B, memory=5)
```

Comment:
maybe rename as `GMRESSolver` ?

Reply:
Regarding the signature, I would prefer we use `;` before specifying the optional argument: 
```julia
function GmresSolver(A, B; memory=5)
```

Reply:
If we use `;`, it not anymore an optional argument but a keyword argument. So you prefer a keyword argument?

Reply:
I would prefer a keyword argument.